contracts/lotterybook/drawer.go

```diff
@@ -0,0 +1,477 @@
+	expired, err := drawer.lmgr.expiredLotteris()
+	if err != nil {
+		return common.Hash{}, err
+	}
+	for _, l := range expired {
+		newId, err := drawer.resetLottery(context, l.Id, payers, amounts, revealNumber)
+		if err != nil {
+			continue
+		}
+		return newId, nil
+	}
+	return drawer.createLottery(context, payers, amounts, revealNumber)
+}
+
+// destroyLottery destroys a stale lottery, claims all deposit inside back to
+// our own pocket.
+func (drawer *ChequeDrawer) destroyLottery(context context.Context, id common.Hash) error {
+	// Short circuit if the specified stale lottery doesn't exist.
+	lottery := drawer.cdb.readLottery(drawer.address, id)
+	if lottery == nil {
+		return errors.New("the lottery specified is not-existent")
+	}
+	// The on-chain transaction may fail dues to lots of reasons like
+	// the lottery doesn't exist or lottery hasn't been expired yet.
+	start := time.Now()
+	tx, err := drawer.book.contract.DestroyLottery(drawer.txSigner, id)
+	if err != nil {
+		return err
+	}
+	receipt, err := bind.WaitMined(context, drawer.dBackend, tx)
+	if err != nil {
+		return err
+	}
+	if receipt.Status != types.ReceiptStatusSuccessful {
+		return ErrTransactionFailed
+	}
+	destroyDurationTimer.UpdateSince(start)
+	// Update manager's status
+	if err := drawer.lmgr.deleteExpired(id); err != nil {
+		return err
+	}
+	// Wipe useless lottery and associated cheques
+	drawer.cdb.deleteLottery(drawer.address, id, false)
+	_, addresses := drawer.cdb.listCheques(
+		drawer.address,
+		func(addr common.Address, lid common.Hash) bool { return lid == id },
+	)
+	for _, addr := range addresses {
+		drawer.cdb.deleteCheque(drawer.address, addr, id, true)
+	}
+	return nil
+}
+
+// Destroy is a wrapper function of destroyLottery. It destorys all
+// expired lotteries and claim back all deposit inside.
+func (drawer *ChequeDrawer) Destroy(context context.Context) error {
+	expired, err := drawer.lmgr.expiredLotteris()
+	if err != nil {
+		return err
+	}
+	reownLotteryGauge.Inc(1)
+	for _, l := range expired {
+		drawer.destroyLottery(context, l.Id)
+	}
+	return nil
+}
+
+// IssueCheque creates a cheque for issuing specified amount for payee.
+//
+// Many active lotteris can be used to create cheque, we use the simplest
+// strategy here for lottery selection: choose a lottery ticket with the
+// most recent expiration date and the remaining amount can cover the amount
+// paid this time.
+func (drawer *ChequeDrawer) IssueCheque(payee common.Address, amount uint64) (*Cheque, error) {
+	lotteris, err := drawer.lmgr.activeLotteris()
+	if err != nil {
+		return nil, err
+	}
+	sort.Sort(LotteryByRevealTime(lotteris))
+	for _, lottery := range lotteris {
+		cheque := drawer.cdb.readCheque(payee, drawer.address, lottery.Id, true)
+		if cheque == nil {
+			continue
+		}
+		if lottery.balance(payee, cheque) >= amount {
+			return drawer.issueCheque(payee, lottery.Id, amount)
+		}
+	}
+	return nil, ErrNotEnoughDeposit // No suitable lottery found for payment
```

Comment:
In the original design, I want to make the lottery management as a Blackbox to users. They don't need to care about selecting which one for payment. So there are only two answers: YES(you can pay) or NO(you can't pay). It can reduce the complicity. 

> you can never really use the last remaining bit of your allowance

Yes, for example, if the allowance of 3 active lotteries is 2,2,2. I want to pay X 3 tokens, then we can never use the last remaining bit. And I agree your idea that **we should somehow combine the cheques together, instead of paying by a single cheque, we can pay several together**. I will think about it how to do it.

Reply:
But maybe I will just leave a todo here(if it involves too many changes :)) ).

Reply:
Done, use multiple lotteries for payment.